crates/aptos-dkg/src/range_proofs/univariate_range_proof.rs

```diff
@@ -542,143 +553,149 @@ pub fn batch_verify(
     );
 
     // Verify h(\tau)
-    let h_check = multi_pairing_g1_g2(
+    let h_check = E::multi_pairing(
         (0..pp.ell)
-            .map(|j| proof.c[j] * betas[j])
-            .chain(once(-proof.d))
-            .collect::<Vec<_>>()
-            .iter(),
+            .map(|j| proof.c[j] * betas[j]) // E::G1
+            .chain(once(-proof.d)) // add -d
+            .collect::<Vec<_>>(), // collect into Vec<E::G1>
         (0..pp.ell)
-            .map(|j| proof.c_hat[j] - pp.taus.t2[0])
-            .chain(once(pp.vanishing_com))
-            .collect::<Vec<_>>()
-            .iter(),
+            .map(|j| proof.c_hat[j] - pp.taus.t2[0]) // E::G2
+            .chain(once(pp.vanishing_com)) // add vanishing commitment
+            .collect::<Vec<_>>(), // collect into Vec<E::G2>
     );
-    ensure!(Gt::identity() == h_check);
+    ensure!(E::TargetField::ONE == h_check.0);
 
     // Ensure duality: c[j] matches c_hat[j].
-    let c_check = multi_pairing_g1_g2(
-        vec![g1_multi_exp(&proof.c, &alphas), -pp.taus.t1[0]].iter(),
-        vec![pp.taus.t2[0], g2_multi_exp(&proof.c_hat, &alphas)].iter(),
+
+    // Compute MSM in G1: sum_j (alphas[j] * proof.c[j])
+    let g1_comb = VariableBaseMSM::msm(
+        &proof
+            .c
+            .iter()
+            .map(|p| p.into_affine())
```

Comment:
I wonder: are we converting to affine from something like projective coordinates just for the MSM code to convert us back?

If so, this wastes performance.